[doc][tiered storage] read data from filesystem (#15239)
apache/pulsar: site2/docs/tiered-storage-filesystem.md

    And the **Capacity Used** is changed from 4 KB to 116.46 KB.

    ![](assets/FileSystem-8.png)

## Read data from fileSystem

This section provides detailed instructions about how to read data out as Ledger Entry in the file system.

* The data was offloaded as `MapFile` to the following path:
  ```properties
    path = storageBasePath + "/" + managedLedgerName + "/" + ledgerId + "-" + uuid.toString();
  ```

>     1. storageBasePath is the value of `hadoop.tmp.dir`, configured in `broker.conf` and `filesystem_offload_core_site.xml`

Reviewer: ```suggestion
    1. storageBasePath is the value of `hadoop.tmp.dir`, configured in `broker.conf` or `filesystem_offload_core_site.xml`
```

    2. managedLedgerName is the name of the persistentTopic manager Ledger
  ```shell

>      managedLedgerName of persistent://public/default/topics-name is  public/default/persistent/topics-name.

Reviewer: ```suggestion
     managedLedgerName of persistent://public/default/topics-name is public/default/persistent/topics-name.
```

  ```

>   Considering the iteration of versions,  you can use the following method to get the managedLedgerName:

Reviewer: ```suggestion
  Can use the following method to get the managedLedgerName:
```

  ```shell
     String managedLedgerName = TopicName.get("persistent://public/default/topics-name").getPersistenceNamingEncoding(); 
  ```

* Create a reader to read `MapFile` according to the above path and the `configuration` of the file system
  ```shell
     MapFile.Reader reader = new MapFile.Reader(new Path(dataFilePath),  configuration); 
  ```
* Read data as `LedgerEntry` from FileSystem.
  ```java
     LongWritable key = new LongWritable();
     BytesWritable value = new BytesWritable();
     key.set(nextExpectedId - 1);
     reader.seek(key);
     reader.next(key, value);
     int length = value.getLength();
     long entryId = key.get();
     ByteBuf buf = PooledByteBufAllocator.DEFAULT.buffer(length, length);
     buf.writeBytes(value.copyBytes());

>      LedgerEntryImpl ledgerEntry = LedgerEntryImpl.create(ledgerId, entryId, length, buf);

Reviewer 2: Just want to know what users want to do when they read the entry? The entry is a serialized message, does they will deserialize the entry to the message by themself?

Author: Great advice. I have added the deserialization steps, please review again when you have time.

    And the **Capacity Used** is changed from 4 KB to 116.46 KB.

    ![](assets/FileSystem-8.png)

> ## Read data from fileSystem

Reviewer 3: ```suggestion
## Read offloaded data from fileSystem
```

> ## Read offloaded data from fileSystem

Reviewer 4: ```suggestion
## Read offloaded data from filesystem
```

> This section provides detailed instructions about how to read data out as Ledger Entry in the file system.

Reviewer 4: This sentence can be moved.

> * The data was offloaded as `MapFile` to the following path:

Reviewer 4: ```suggestion
The data is offloaded to the following new path as `MapFile`:
```

  ```properties
    path = storageBasePath + "/" + managedLedgerName + "/" + ledgerId + "-" + uuid.toString();
  ```

>     1. storageBasePath is the value of `hadoop.tmp.dir`, configured in `broker.conf` or `filesystem_offload_core_site.xml`

Reviewer 4: ```suggestion
* `storageBasePath` is the value of `hadoop.tmp.dir`, which is configured in `broker.conf` or `filesystem_offload_core_site.xml`.
```

>     2. managedLedgerName is the name of the persistentTopic manager Ledger

Reviewer 4: * `managedLedgerName` is the ledger name of the persistentTopic manager.

  ```shell
     managedLedgerName of persistent://public/default/topics-name is public/default/persistent/topics-name.
  ```

>   Can use the following method to get the managedLedgerName:

Reviewer 4: ```suggestion
  You can use the following method to get `managedLedgerName`:
```

  ```shell
     String managedLedgerName = TopicName.get("persistent://public/default/topics-name").getPersistenceNamingEncoding(); 
  ```

> * Create a reader to read `MapFile` according to the above path and the `configuration` of the file system

Reviewer 4: ```suggestion
1. Create a reader to read both `MapFile` with a new path and the `configuration` of the filesystem.
```

  ```shell
     MapFile.Reader reader = new MapFile.Reader(new Path(dataFilePath),  configuration); 
  ```

> * Read data as `LedgerEntry` from FileSystem.

Reviewer 4: ```suggestion
2. Read the data as `LedgerEntry` from the filesystem.
```

  ```java
     LongWritable key = new LongWritable();
     BytesWritable value = new BytesWritable();
     key.set(nextExpectedId - 1);
     reader.seek(key);
     reader.next(key, value);
     int length = value.getLength();
     long entryId = key.get();
     ByteBuf buf = PooledByteBufAllocator.DEFAULT.buffer(length, length);
     buf.writeBytes(value.copyBytes());
     LedgerEntryImpl ledgerEntry = LedgerEntryImpl.create(ledgerId, entryId, length, buf);
  ```

> * Deserialize the `ledgerEntry` to `Message`.

Reviewer 4: ```suggestion
3. Deserialize the `ledgerEntry` to `Message`.
```

Which one is correct? `LedgerEntry` or `ledgerEntry`? The difference between step2 and step3 may make users confused.
